src/dataflow-types/src/client.rs

```diff
@@ -779,10 +912,57 @@ pub mod partitioned {
                         None
                     }
                 }
-                message => {
-                    // TimestampBindings and TailResponses are mirrored out,
-                    // as they do not seem to contain worker-specific information.
-                    Some(message)
+                Response::Compute(ComputeResponse::TailResponse(id, response)) => {
+                    let maybe_entry = self.pending_tails.entry(id).or_insert_with(|| {
+                        Some(PendingTail {
+                            frontiers: HashMap::new(),
+                            buffer: Vec::new(),
+                            parts: self.parts,
+                            reported_frontier: Antichain::from_elem(0),
+                        })
+                    });
+
+                    let entry = match maybe_entry {
+                        None => {
+                            // This tail has been dropped;
+                            // we should pernanently block
+                            // any messages from it
+                            return None;
+                        }
+                        Some(entry) => entry,
+                    };
+
+                    let tail_response = match response {
+                        TailResponse::Progress(frontier) => {
+                            if let Some(new_frontier) = entry.record_progress(shard_id, frontier) {
+                                let data = entry.consolidate_up_to(new_frontier.clone());
+                                let progress_response = TailResponse::Progress(new_frontier.into());
+                                if data.is_empty() {
+                                    Some(progress_response)
+                                } else {
+                                    // Return the data first, then the progress message.
+                                    assert!(self.stashed_response.is_none());
+                                    self.stashed_response = Some(Response::Compute(
+                                        ComputeResponse::TailResponse(id, progress_response),
+                                    ));
+                                    Some(TailResponse::Rows(data))
```

Comment:
Musing whether there's a more elegant way to express that a call to `recv` could result in more than one object to be returned. It seems .. we could express this as a stream?